Add focus and blur events that always bubble up all the way to window for autofill implementation

rniwa opened the pull request:
#### d98000f4188c83559b3f3f17667d705afa122a76
<pre>
Add focus and blur events that always bubble up all the way to window for autofill implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=303675">https://bugs.webkit.org/show_bug.cgi?id=303675</a>

Reviewed by Ben Nham.

This PR introduces webkitbeforefocus and webkitbeforeblur events which are equivalent of focus and blur
events but without a related target. The lack of related target causes these events to bubble up all
the way to document node, allowing autofill implementation to efficiently listen to these events without
adding an event listener to each shadow root.

Test: TestWebKitAPI.WKUserContentController.BeforeFocusEvent
      TestWebKitAPI.WKUserContentController.BeforeBlurEvent

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchFocusEvent):
(WebCore::Element::dispatchBlurEvent):
* Source/WebCore/dom/EventNames.json:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, BeforeFocusEvent)):
(TEST(WKUserContentController, BeforeBlurEvent)):

Canonical link: <a href="https://commits.webkit.org/304112@main">https://commits.webkit.org/304112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f476e6915a74a06c33cfb13da693444942137faa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134593 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142118 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86547 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/81f63b87-a5b7-4858-96dc-4bd3ed134ae5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136463 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102875 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70159 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9474d3bf-2558-49c0-8c1c-a751efd863d6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137540 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83675 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d8345370-f881-4a76-9606-4421133a0e5d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5219 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2838 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2716 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38781 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144811 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6726 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39363 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111276 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111562 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5056 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116915 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60613 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20785 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6776 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35107 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6587 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70360 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6823 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6698 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4291,19 +4291,28 @@ void Element::dispatchFocusEvent(RefPtr<Element>&& oldFocusedElement, const Focu
 
 void Element::dispatchBlurEvent(RefPtr<Element>&& newFocusedElement)
 {
+    Ref beforefocusEvent = Event::create(eventNames().webkitbeforeblurEvent, Event::CanBubble::Yes, Event::IsCancelable::No, Event::IsComposed::Yes);
```
nit: `Ref beforeblurEvent =`

rniwa committed:
Add focus and blur events that always bubble up all the way to window for autofill implementation
https://bugs.webkit.org/show_bug.cgi?id=303675

Reviewed by Ben Nham.

This PR introduces webkitbeforefocus and webkitbeforeblur events which are equivalent of focus and blur
events but without a related target. The lack of related target causes these events to bubble up all
the way to document node, allowing autofill implementation to efficiently listen to these events without
adding an event listener to each shadow root.

Test: TestWebKitAPI.WKUserContentController.BeforeFocusEvent
      TestWebKitAPI.WKUserContentController.BeforeBlurEvent

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchFocusEvent):
(WebCore::Element::dispatchBlurEvent):
* Source/WebCore/dom/EventNames.json:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, BeforeFocusEvent)):
(TEST(WKUserContentController, BeforeBlurEvent)):

Canonical link: https://commits.webkit.org/304112@main
